Bugfix for RPNV1-148

motoras opened the pull request:
Switched to trace level logging with lambda parameters in BFTEventReducer.
Logging starting view message in ScheduledTimeoutSender only after at least one second had passed since the previous log statement.

For ticket details see: https://radixdlt.atlassian.net/browse/RPNV1-148

Commit:
Log statements in BFTEventReducer use TRACE level.
Create view log message in ScheduledTimeoutSender is printed only after
a given time interval.

Commit:
Use volatile timestamp to check logging time interval.

Commit:
Extra whitespaces removed.

Commit:
Use lambda for logging.

clstrfsck commented on radixdlt/src/main/java/com/radixdlt/consensus/liveness/ScheduledTimeoutSender.java:
```diff
@@ -49,7 +51,13 @@ public ScheduledTimeoutSender(ScheduledExecutorService executorService) {
 	@Override
 	public void scheduleTimeout(final View view, long timeoutMilliseconds) {
-		log.info("Starting View: {}", view);
+	    long crtTime = System.currentTimeMillis();
```
Looks like editor is using spaces instead of tabs here.

motoras replied:
Somehow, this method still uses space instead of tabs for alignment. I'll fix it.

clstrfsck commented on radixdlt/src/main/java/com/radixdlt/consensus/liveness/ScheduledTimeoutSender.java:
```diff
@@ -17,6 +17,7 @@
 
 package com.radixdlt.consensus.liveness;
 
+
```
Extra blank line.

clstrfsck commented on radixdlt/src/main/java/com/radixdlt/consensus/BFTEventReducer.java:
```diff
@@ -234,18 +232,16 @@ public void processProposal(Proposal proposal) {
 			this.proceedToView(nextView.get());
-			log.info("{}: LOCAL_TIMEOUT: Processed {}", this.getShortName(), view);
-
+			log.trace("{}: LOCAL_TIMEOUT: Processed {}", this::getShortName, () -> view);
```
I think that this should be processed at INFO or at least DEBUG level, as this indicates a timeout has occurred, which should be a reasonably unusual event.

motoras replied:
I'll change it to warning.

clstrfsck commented on radixdlt/src/main/java/com/radixdlt/consensus/BFTEventReducer.java:
```diff
@@ -201,9 +202,7 @@ public void processProposal(Proposal proposal) {
-
-			log.info(String.format("%s: PROPOSAL: Rejected", this.getShortName()), e);
-
+			log.trace("{} PROPOSAL: Rejected. Reason: {}", this::getShortName, e::getMessage);
```
I think this should be logged at INFO level at least.  This should be a quite unusual event in the final system, and indicates that some attention is really required.

clstrfsck replied:
Probably should be a warning.

motoras replied:
I agree

clstrfsck commented on radixdlt/src/main/java/com/radixdlt/consensus/BFTEventReducer.java:
```diff
@@ -153,43 +154,43 @@ public void processLocalSync(Hash vertexId) {
 		this.pendingVotes.insertVote(vote, this.validatorSet).ifPresent(qc -> {
-			log.info("{}: VOTE: Formed QC: {}", this.getShortName(), qc);
+			log.trace("{}: VOTE: Formed QC: {}", this::getShortName, () -> qc);
 			if (vertexStore.syncToQC(qc, vertexStore.getHighestCommittedQC(), vote.getAuthor())) {
```
I think it would be really convenient to see when the node becomes unsynched, and then resynched here.  Something like:
```
	if (vertexStore.syncToQC(qc, vertexStore.getHighestCommittedQC(), vote.getAuthor())) {
		if (!synchedLog) {
			log.info("{}: VOTE: QC Synced: {}", this::getShortName, () -> qc);
			synchedLog = true;
		}
		processQC(qc);
	} else {
		if (synchedLog) {
			log.info("{}: VOTE: QC Not synced: {}", this::getShortName(), () -> qc);
			synchedLog = false;
		}
		unsyncedQCs.put(qc.getProposed().getId(), qc);
	}
```
...so we can track how frequently nodes become synced and unsynced in logs at a reasonably high log level, without flooding logs.

motoras replied:
Shall we add the same logging to processLocalSync?

clstrfsck replied:
Could do, yes.

clstrfsck replied:
You could if you wanted to.  I don't think it's so important there, and probably shouldn't be at the info level if it goes in.

flotothemoon commented on radixdlt/src/main/java/com/radixdlt/consensus/liveness/ScheduledTimeoutSender.java:
```diff
@@ -49,7 +51,13 @@ public ScheduledTimeoutSender(ScheduledExecutorService executorService) {
 	@Override
 	public void scheduleTimeout(final View view, long timeoutMilliseconds) {
-		log.info("Starting View: {}", view);
+	    long crtTime = System.currentTimeMillis();
```
Formatting (indent) seems off here

clstrfsck reviewed (changes requested):
Some suggested changes that will improve logging without necessarily increasing log size significantly.

flotothemoon reviewed (commented):
I'm not convinced that the majority of log messages affected here are really "trace" level messages. They seem to be important information to me. Their tendency to spam when busy spinning could be reduced in other ways (e.g. filter out logging of the specific class)

clstrfsck commented:
Note also that the target branch is incorrect.  It should be targetting `rc/1.0-beta.11`.

talekhinezh commented on radixdlt/src/main/java/com/radixdlt/consensus/liveness/ScheduledTimeoutSender.java:
```diff
@@ -33,10 +34,11 @@
 	private final ScheduledExecutorService executorService;
 	private final Subject<View> timeouts;
 	private final Observable<View> timeoutsObservable;
+	private volatile long nextLogging = 0;
```
Doesn't need to be volatile as this class should only ever be accessed by a single thread.

motoras replied:
ok

clstrfsck commented:
> I'm not convinced that the majority of log messages affected here are really "trace" level messages. They seem to be important information to me. Their tendency to spam when busy spinning could be reduced in other ways (e.g. filter out logging of the specific class)

Hmm, maybe.  But I think that argument works the other way too.  If the logs are important (for whatever reason), then the level can be increased for the class to include them.

Maybe logging them at the lowest level is not ideal, and we should consider making them DEBUG instead of TRACE level?

Important to remember here that logs in general, but particularly at INFO level and above should really be targeted at operational management.

Commit:
BFTEventReducer logging and code formatting improved.
Removed the volatile attribute of the nextLogging field in ScheduledTimeoutSender.

Commit:
Imports statements ordered correctly.

clstrfsck commented on radixdlt/src/main/java/com/radixdlt/consensus/BFTEventReducer.java:
```diff
@@ -64,21 +63,22 @@
-		ValidatorSet validatorSet,
-		SystemCounters counters
-	) {
+			ProposalGenerator proposalGenerator,
```
Oof.  Can you change this reformat back to what it was please?

clstrfsck commented on radixdlt/src/main/java/com/radixdlt/consensus/BFTEventReducer.java:
```diff
@@ -105,46 +105,45 @@ private void proceedToView(View nextView) {
-			this.vertexStore.getHighestCommittedQC(),
-			signature
-		);
+				selfKey.getPublicKey(),
```
Can you change this reformat back to what it was please?

clstrfsck commented on radixdlt/src/main/java/com/radixdlt/consensus/BFTEventReducer.java:
```diff
@@ -105,46 +105,45 @@ private void proceedToView(View nextView) {
-					}
-				});
-
+		.ifPresent(commitMetaData -> {
```
Can you change this reformat back to what it was please?

motoras replied:
Sure!

Commit:
Identation fixed.

clstrfsck reviewed (approved):
Looks good to me.  I think the additional unit tests to keep sonarcloud happy are not worth the effort here.
